File: assertions/check_technical_accuracy.ts
/**
 * TypeScript Assertion: Prüft technische Korrektheit
 * 
 * Usage in Promptfoo:
 *   assert:
 *     - type: javascript
 *       value: file://assertions/check_technical_accuracy.ts
 */

interface AssertionResult {
  pass: boolean;
  score: number;
  reason: string;
}

interface AssertionContext {
  vars: Record<string, string>;
  prompt: string;
}

// Technische Fakten die korrekt sein müssen
const TECHNICAL_FACTS: Record<string, { correct: string[]; incorrect: string[] }> = {
  vin_length: {
    correct: ['17 zeichen', '17 stellen', '17-stellig', 'siebzehn zeichen', 'siebzehn stellen'],
    incorrect: ['16 zeichen', '18 zeichen', '15 zeichen', '20 zeichen']
  },
  vin_forbidden_chars: {
    correct: ['i, o, q', 'i o q', 'buchstaben i', 'buchstabe o', 'buchstabe q'],
    incorrect: ['alle buchstaben erlaubt', 'keine einschränkungen']
  },
  wmi_meaning: {
    correct: ['world manufacturer identifier', 'hersteller', 'herstellerkennung'],
    incorrect: ['world motor identifier', 'world model identifier']
  },
  euro_6d: {
    correct: ['abgasnorm', 'emission', 'schadstoff', 'nox'],
    incorrect: ['kraftstoffnorm', 'sicherheitsnorm']
  },
  adblue: {
    correct: ['harnstoff', 'scr', 'nox', '32,5%', '32.5%'],
    incorrect: ['diesel', 'benzin', 'kühlmittel']
  }
};

function getAssert(output: string, context: AssertionContext): AssertionResult {
  if (!output || typeof output !== 'string') {
    return {
      pass: false,
      score: 0,
      reason: 'Keine Ausgabe vorhanden'
    };
  }

  const outputLower = output.toLowerCase();
  const query = context.vars?.query?.toLowerCase() || '';
  
  const issues: string[] = [];
  const correctFacts: string[] = [];
  let relevantChecks = 0;
  let passedChecks = 0;

  // Bestimme relevante Fakten basierend auf der Frage
  const relevantTopics: string[] = [];
  
  if (query.includes('vin') || query.includes('fin') || query.includes('fahrzeug-identifikation')) {
    relevantTopics.push('vin_length', 'vin_forbidden_chars', 'wmi_meaning');
  }
  if (query.includes('euro') || query.includes('abgas')) {
    relevantTopics.push('euro_6d');
  }
  if (query.includes('adblue') || query.includes('def') || query.includes('harnstoff')) {
    relevantTopics.push('adblue');
  }

  // Wenn keine spezifischen Topics, prüfe alle
  const topicsToCheck = relevantTopics.length > 0 ? relevantTopics : Object.keys(TECHNICAL_FACTS);

  for (const topic of topicsToCheck) {
    const facts = TECHNICAL_FACTS[topic];
    if (!facts) continue;

    // Prüfe ob Topic im Output erwähnt wird
    const topicMentioned = facts.correct.some(c => outputLower.includes(c)) ||
                          facts.incorrect.some(i => outputLower.includes(i));

    if (topicMentioned) {
      relevantChecks++;

      // Prüfe auf inkorrekte Fakten
      const hasIncorrect = facts.incorrect.some(i => outputLower.includes(i));
      const hasCorrect = facts.correct.some(c => outputLower.includes(c));

      if (hasIncorrect && !hasCorrect) {
        issues.push(`Inkorrekt bei ${topic}`);
      } else if (hasCorrect) {
        passedChecks++;
        correctFacts.push(topic);
      }
    }
  }

  // Berechne Score
  const score = relevantChecks > 0 ? passedChecks / relevantChecks : 1;
  const pass = issues.length === 0 && score >= 0.8;

  return {
    pass,
    score,
    reason: pass
      ? `Technisch korrekt (${correctFacts.length} Fakten geprüft)`
      : `Technische Fehler: ${issues.join(', ')}`
  };
}

// CommonJS Export für Promptfoo
module.exports = getAssert;
module.exports.getAssert = getAssert;
module.exports.TECHNICAL_FACTS = TECHNICAL_FACTS;

File: assertions/check_technical_accuracy.py
"""
Custom Python Assertion: Prüft technische Genauigkeit für LKW-Informationen
Verwendung in promptfooconfig.yaml:
  assert:
    - type: python
      value: file://assertions/check_technical_accuracy.py
"""

import re

# Bekannte technische Fakten
TECHNICAL_FACTS = {
    "vin_length": 17,
    "wmi_length": 3,
    "forbidden_vin_chars": ["I", "O", "Q"],
    "man_wmi_codes": ["WMA", "WMH", "XMC"],
    "euro_norms": ["Euro 5", "Euro 6", "Euro 6c", "Euro 6d", "Euro 6e"],
    "man_models": ["TGL", "TGM", "TGS", "TGX", "eTGM", "eTGS", "eTGX", "Lion's City"],
    "axle_configs": ["4x2", "4x4", "6x2", "6x4", "8x4", "8x8"],
}


def get_assert(output: str, context: dict) -> dict:
    """
    Prüft technische Genauigkeit der Antwort.
    
    Args:
        output: Die LLM-Antwort
        context: Kontext mit vars, prompt, etc.
    
    Returns:
        dict mit pass, score, reason
    """
    errors = []
    warnings = []
    
    output_lower = output.lower()
    
    # Prüfe VIN-Länge wenn erwähnt
    vin_length_match = re.search(r'vin.*?(\d+)[\s-]*(stellig|zeichen|character)', output_lower)
    if vin_length_match:
        mentioned_length = int(vin_length_match.group(1))
        if mentioned_length != TECHNICAL_FACTS["vin_length"]:
            errors.append(f"VIN-Länge falsch: {mentioned_length} statt 17")
    
    # Prüfe WMI-Länge wenn erwähnt
    wmi_length_match = re.search(r'wmi.*?(\d+)[\s-]*(stellig|zeichen|stellen)', output_lower)
    if wmi_length_match:
        mentioned_length = int(wmi_length_match.group(1))
        if mentioned_length != TECHNICAL_FACTS["wmi_length"]:
            errors.append(f"WMI-Länge falsch: {mentioned_length} statt 3")
    
    # Prüfe ob verbotene VIN-Zeichen korrekt genannt werden
    if "vin" in output_lower and ("buchstabe" in output_lower or "zeichen" in output_lower):
        if "i" in output_lower and "o" in output_lower:
            # Gut - beide verbotenen Buchstaben erwähnt
            pass
        elif "verwechsl" in output_lower:
            # Erwähnt Verwechslungsgefahr - auch OK
            pass
    
    # Prüfe Euro-Normen
    for norm in ["euro 7", "euro 8"]:
        if norm in output_lower:
            warnings.append(f"'{norm}' erwähnt - existiert noch nicht für LKW")
    
    # Prüfe auf offensichtlich falsche Zahlen
    # z.B. "VIN hat 15 Zeichen" oder "WMI hat 5 Stellen"
    
    # Berechne Score
    if errors:
        score = 0.0
        passed = False
        reason = f"Technische Fehler: {'; '.join(errors)}"
    elif warnings:
        score = 0.7
        passed = True
        reason = f"Technisch korrekt mit Warnungen: {'; '.join(warnings)}"
    else:
        score = 1.0
        passed = True
        reason = "Technisch korrekt"
    
    return {
        "pass": passed,
        "score": score,
        "reason": reason
    }
